cachi2/core/resolver.py

```diff
@@ -14,21 +14,27 @@
     "pip": pip.fetch_pip_source,
 }
 
+_dev_package_managers: dict[PackageManagerType, Handler] = {
+    "yarn": yarn.fetch_yarn_source,
+}
 
 supported_package_managers = list(_package_managers)
 
 
 def resolve_packages(request: Request) -> RequestOutput:
     """Run all requested package managers, return their combined output."""
+    local_pkg_managers = _package_managers
     requested_types = set(pkg.type for pkg in request.packages)
-    unsupported_types = requested_types - _package_managers.keys()
+    if "dev-mode" in request.flags:
+        local_pkg_managers |= _dev_package_managers
```

Comment:
I still think we shouldn't update the global `_package_managers` variable, and use a local copy. Although Cachi2 is very likely to be used as a "single run" process, I'd avoid this as it could become a pitfall in some alternate scenarios.

Reply:
`local_pkg_managers` **should** be a (true) _copy_ of `_package_managers`... If it weren't, that would defeat the entire purpose of the `global` keyword.
I'll run a debug and check object addresses to be certain.

Reply:
If we only assign a new variable like `local_pkg_managers = _package_managers`, we're actually just passing the reference to the same dictionary to another variable, so no actual copy is made.

Reply:
Correct, but in that scope, `_package_managers` does *not* point to the module-level "global" `_package_managers` - it's a copy

Reply:
As weird as it seems, it does point to the module-level variable :)

You can see it if you print the module-level variable after the `|=` operation, and make a request with `--dev-mode`.
```
diff --git a/cachi2/core/resolver.py b/cachi2/core/resolver.py
index 664332b..5bc61d3 100644
--- a/cachi2/core/resolver.py
+++ b/cachi2/core/resolver.py
@@ -34,6 +34,7 @@ def resolve_packages(request: Request) -> RequestOutput:
             # unknown package managers shouldn't get past input validation
             solution="But the good news is that we're already working on it!",
         )
+    print(_package_managers)
     pkg_managers = [local_pkg_managers[type_] for type_ in sorted(requested_types)]
     return _merge_outputs(pkg_manager(request) for pkg_manager in pkg_managers)
```

Reply:
I know, right?? I usually do a lot of small tests locally to remember how these details work.. it never sticks to my head :shrug:

Reply:
Ah. Apparently it's in the fine print: "[free variables may refer to globals without being declared global](https://docs.python.org/3.9/reference/simple_stmts.html#the-global-statement)", although I'm sure that I don't understand all of the implications of that statement.